columnflow/calibration/cms/egamma.py

```diff
@@ -0,0 +1,581 @@
+# coding: utf-8
+
+"""
+Egamma energy correction methods.
+Source: https://twiki.cern.ch/twiki/bin/view/CMS/EgammSFandSSRun3#Scale_And_Smearings_Correctionli
+"""
+
+from __future__ import annotations
+
+import abc
+import functools
+import itertools
+import law
+from dataclasses import dataclass, field
+
+from columnflow.calibration import Calibrator, calibrator
+from columnflow.calibration.util import ak_random
+from columnflow.util import maybe_import, InsertableDict
+from columnflow.columnar_util import (
+    set_ak_column, flat_np_view, ak_copy, optional_column,
+)
+from columnflow.types import Any
+
+ak = maybe_import("awkward")
+np = maybe_import("numpy")
+
+
+# helper
+set_ak_column_f32 = functools.partial(set_ak_column, value_type=np.float32)
+
+
+@dataclass
+class EGammaCorrectionConfig:
+    correction_set: str = "Scale"
+    corrector_kwargs: dict[str, Any] = field(default_factory=dict)
+
+
+class egamma_scale_corrector(Calibrator):
+
+    with_uncertainties = True
+    """Switch to control whether uncertainties are calculated."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    @property
+    @abc.abstractmethod
+    def source_field(self) -> str:
+        """Fields required for the current calibrator."""
+        ...
+
+    @abc.abstractmethod
+    def get_correction_file(self, external_files: law.FileTargetCollection) -> law.LocalFile:
+        """Function to retrieve the correction file from the external files.
+
+        :param external_files: File target containing the files as requested
+            in the current config instance under ``config_inst.x.external_files``
+        """
+        ...
+
+    @abc.abstractmethod
+    def get_scale_config(self) -> EGammaCorrectionConfig:
+        """Function to retrieve the configuration for the photon energy correction."""
+        ...
+
+    def call_func(
+        self: Calibrator,
```

Comment:
Same comment as for the seed producers below, we usually don't need to have type hints for `self` (same below)